core/process.py

```diff
@@ -1499,7 +1519,13 @@ def WhenDone(self, pid):
         if pid == self.last_stopped_pid:
             self.last_stopped_pid = -1
 
-        mylib.dict_erase(self.jobs, pid)
+        assert pid in self.child_procs
+        proc = self.child_procs[pid]
+        if proc.job_id in self.jobs:
+            mylib.dict_erase(self.jobs, proc.job_id)
+            # not all background processes (e.g. command subs, subshells) have jobs
+            if not proc.InForeground():
```

Comment:
Hm  is this the only thing we're using the flag for?

How about:

1. we add a flag to Process and Pipeline `self.in_background = False` 
2. There are 2 functions in the `_Executor` where we start processes and pipelines in the background.  In those cases we do `obj.SetBackground()`

I don't really like "MoveToForeground() MoveToBackground()" because it sounds like some syscalls are  happening

Reply:
Ah yeah, `Move*()` is a little misleading. Will use the new naming